File: parser/service.go
package parser

import (
	"fmt"
	"strings"
)

func GetService(data string) string {
	var serviceName string
	for _, line := range strings.Split(data, "\n") {
		if strings.Contains(line, "service ") {
			serviceName = strings.Split(line, " ")[1]
		}
		if strings.Contains(line, "rpc ") {
			rpcParser(line)
		}
	}
	fmt.Println("service name : ", serviceName)
	return serviceName
}

func rpcParser(data string) {
	data = strings.Replace(data, "{}", "", -1)
	data = strings.Replace(data, ";", "", -1)
	data = strings.Replace(data, "rpc", "", -1)
	data = strings.Replace(data, "returns", "", -1)
	data = strings.TrimSpace(data)
	splitter := strings.Split(data, ")  (")
	prefix := strings.Split(splitter[0], "(")[0]
	requestParams := strings.Split(splitter[0], "(")[1]
	requestParams = strings.Replace(requestParams, ")", "", -1)
	responseParam := strings.Split(splitter[1], ")")[0]
	fmt.Println("rpc prefix : ", prefix)
	fmt.Println("rpc request params : ", requestParams)
	fmt.Println("rpc response params : ", responseParam)
	fmt.Println("========================================")

}

File: parser/service.py
def create_service(repo_name : str ,data: dict):
    print("Creating service...")
    service_name: str = data.get("service")
    service_data = data.get("service_data")
    service_name = service_name.replace("Service", "")
    template = """
package service

import (
    "context"
    "fmt"
 
    "gorm.io/gorm"
)

type {service_name}Service struct {{
    pb.UnimplementedLoanServiceServer
	Db *gorm.DB
}}
""".format(service_name=service_name)

    for item in service_data:
        name = item.get("name")
        request = item.get("request")
        response = item.get("response")
        
        func = """
func (s *{service_name}Service) {name}(ctx context.Context, param *pb.{request_name}) (*pb.{response_name}, error) {{
    repo := repository.Init{service_name}Repository(s.Db)
    response, err := repo.{name}(param)
    if err != nil {{
        return nil, err
    }}
    return resp, nil
}}
        """.format(service_name=service_name, name=name, request_name=request, response_name=response)
        template = template + func
    
    with open(f"result/rpc/{service_name}.service.go", "w") as f:
        f.write(template)
    print(f"Service '{service_name}' created.")
